### PR TITLE
[FSTORE-921] Resolve serving key in feature view

### PR DESCRIPTION
**Pull request by kennethmhc**
- add `allow_missing` to `get_feature_vector(s)` and return feature vector with missing value 
- fix the logic of required primary key (serving key)

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/core/vector_server.py [hunk 1]
```diff
@@ -32,14 +33,22 @@
...
+    def __init__(
+        self,
+        feature_store_id,
+        features=[],
+        training_dataset_version=None,
+        serving_keys=None,
+    ):
         self._training_dataset_version = training_dataset_version
         self._features = features
+        self._feature_vector_col = [feat for feat in features if not feat.label]
```
Comment (davitbzh) [inline python/hsfs/core/vector_server.py:45]: `fv.create_training_data()` any flavor of it throws error:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-18-8940fa0ea0cb> in <module>
----> 1 fv1.create_training_data()

/srv/hops/anaconda/envs/theenv/lib/python3.10/site-packages/hsfs/feature_view.py in create_training_data(self, start_time, end_time, storage_connector, location, description, extra_filter, data_format, coalesce, seed, statistics_config, write_options, spine)
    872                 that was launched to create the training dataset.
    873         """
--> 874         td = training_dataset.TrainingDataset(
    875             name=self.name,
    876             version=None,

/srv/hops/anaconda/envs/theenv/lib/python3.10/site-packages/hsfs/training_dataset.py in __init__(self, name, version, data_format, featurestore_id, location, event_start_time, event_end_time, coalesce, description, storage_connector, splits, validation_size, test_size, train_start, train_end, validation_start, validation_end, test_start, test_end, seed, created, creator, features, statistics_config, featurestore_name, id, inode_id, training_dataset_type, from_query, querydto, label, transformation_functions, train_split, time_split_size, extra_filter)
    182             self._label = [feat.name.lower() for feat in self._features if feat.label]
    183             self._extra_filter = filter.Logic.from_response_json(extra_filter)
--> 184         self._vector_server = vector_server.VectorServer(
    185             featurestore_id, features=self._features
    186         )

/srv/hops/anaconda/envs/theenv/lib/python3.10/site-packages/hsfs/core/vector_server.py in __init__(self, feature_store_id, features, training_dataset_version, serving_keys)
     43         self._training_dataset_version = training_dataset_version
     44         self._features = features
---> 45         self._feature_vector_col = [feat for feat in features if not feat.label]
     46         self._prepared_statement_engine = None
     47         self._prepared_statements = None

TypeError: 'NoneType' object is not iterable

File: python/hsfs/core/vector_server.py [hunk 2]
```diff
@@ -480,7 +495,10 @@ def prepared_statements(self, prepared_statements):
     @property
     def serving_keys(self):
         """Set of primary key names that is used as keys in input dict object for `get_serving_vector` method."""
```
Comment (davitbzh) [inline python/hsfs/core/vector_server.py:497]: This is very minor but I guess its better to have `get_feature_vector`

File: python/hsfs/core/vector_server.py [hunk 3]
```diff
@@ -156,117 +192,118 @@ def get_feature_vector(self, entry, return_type, passed_features={}):
...
+        vector = self._generate_vector(result_dict)
 
         if return_type.lower() == "list":
             return vector
         elif return_type.lower() == "numpy":
             return np.array(vector)
         elif return_type.lower() == "pandas":
             pandas_df = pd.DataFrame(vector).transpose()
-            pandas_df.columns = feature_names
+            pandas_df.columns = self._feature_vector_col
```
Comment (davitbzh) [inline python/hsfs/core/vector_server.py:238]: pandas returns strange column names:
<html><head></head><body>

Training Dataset Feature('id1', 'bigint', 0, False, None, None, 14) | Training Dataset Feature('fg1_id1', 'bigint', 1, False, None, None, 15) | Training Dataset Feature('fg1_id2', 'bigint', 2, False, None, None, 15)
-- | -- | --
6 | 6 | 7

</body></html>

File: python/hsfs/core/vector_server.py [hunk 4]
```diff
@@ -301,7 +330,7 @@ def get_feature_vectors(self, entry, return_type, passed_features=[]):
             return np.array(vectors)
         elif return_type.lower() == "pandas":
             pandas_df = pd.DataFrame(vectors)
-            pandas_df.columns = feature_names
+            pandas_df.columns = self._feature_vector_col
```
Comment (davitbzh) [inline python/hsfs/core/vector_server.py:333]: The same here, column names are strange:
<html><head></head><body>

  | Training Dataset Feature('id1', 'bigint', 0, False, None, None, 14) | Training Dataset Feature('fg1_id1', 'bigint', 1, False, None, None, 15) | Training Dataset Feature('fg1_id2', 'bigint', 2, False, None, None, 15)
-- | -- | -- | --
7 | 7 | 8
6 | 6 | 7

</body></html>

**Review by davitbzh — CHANGES_REQUESTED** [review]
See my comments above. Also would be great to add examples how to fetch primary keys in `get_feature_vector` and `get_feature_vectors` docstring examples.

Also I think 1 notebook in load test will be great. You can add the same notebook you sent me

**Comment by kennethmhc** [comment]
@davitbzh addressed your comment. I will make a PR to the load test.

File: python/hsfs/feature_view.py [hunk 5]
```diff
@@ -208,6 +211,10 @@ def init_serving(
                   For example: `{"pool_size": 10}`
         """
 
+        # initiate batch scoring server
+        # `training_dataset_version` should not be set if `None` otherwise backend will look up the td.
+        self.init_batch_scoring(training_dataset_version)
```
Comment (kennethmhc) [inline python/hsfs/feature_view.py:216]: bug fix

**Review by davitbzh — APPROVED** [review]

**Pull request merged**